components/operations-controller/controllers/operation_controller.go

```diff
@@ -16,36 +16,420 @@ package controllers
+	if err != nil {
+		logger.Error(err, operationError("Unable to fetch application", operation))
+
+		if isReconciliationTimeoutReached(operation, time.Duration(r.webhookConfig.TimeoutFactor)*r.webhookConfig.ReconciliationTimeout) {
+			if err := r.k8sClient.Delete(ctx, operation); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{RequeueAfter: r.webhookConfig.RequeueInterval}, nil
+	}
+
+	operation = setDefaultStatus(*operation)
+
+	if app.Result.Ready {
+		operation = setConditionStatus(*operation, v1alpha1.ConditionTypeReady, corev1.ConditionTrue, "")
+		operation.Status.Phase = v1alpha1.StateSuccess
+		if app.Result.Error != nil && *app.Result.Error != "" {
+			operation = setConditionStatus(*operation, v1alpha1.ConditionTypeError, corev1.ConditionTrue, *app.Result.Error)
+			operation.Status.Phase = v1alpha1.StateFailed
+		}
+		if err := r.k8sClient.UpdateStatus(ctx, operation); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if app.Result.Error != nil && *app.Result.Error != "" {
+		operation = setErrorStatus(*operation, *app.Result.Error)
+		operation.Status.Phase = v1alpha1.StateFailed
+		if err := r.k8sClient.UpdateStatus(ctx, operation); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	webhooks, err := retrieveWebhooks(app.Result.Webhooks, operation.Spec.WebhookIDs)
+	if err != nil {
+		logger.Error(err, operationError("Unable to retrieve webhooks", operation))
+		return ctrl.Result{}, nil
+	}
+
+	if isReconciliationTimeoutReached(operation, calculateReconciliationTimeout(webhooks, r.webhookConfig.WebhookTimeout)) {
+		opErr := errors.New("reconciliation timeout reached")
+		return r.updateStatusWithError(ctx, logger, operation, opErr)
+	}
+
+	webhookEntity := webhooks[operation.Spec.WebhookIDs[0]]
+	webhookStatus := &operation.Status.Webhooks[0]
+
+	if webhookStatus.WebhookPollURL == "" {
+		request := web_hook.NewRequest(webhookEntity, requestData, operation.Spec.CorrelationID)
+
+		response, err := r.webhookClient.Do(ctx, request)
+		if err != nil {
+			logger.Error(err, operationError("Unable to execute Webhook request", operation))
+			return r.handleWebhookTimeoutCheckResponse(ctx, logger, operation, webhookEntity, err)
+		}
+
+		switch *webhookEntity.Mode {
+		case graphql.WebhookModeAsync:
+			operation.Status.Webhooks[0].WebhookPollURL = *response.Location
+			if err := r.k8sClient.UpdateStatus(ctx, operation); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{Requeue: true}, nil
+		case graphql.WebhookModeSync:
+			return r.updateStatus(ctx, logger, operation)
+		default:
+			return ctrl.Result{}, errors.New("unsupported webhook mode")
+		}
+	}
+
+	requeueAfter, err := getNextPollTime(&webhookEntity, *webhookStatus, r.webhookConfig.TimeLayout)
+	if err != nil {
+		logger.Error(err, operationError("Unable to calculate next poll time", operation))
```

Comment:
The only reason for the `getNextPollTime` to fail is if the `r.webhookConfig.TimeLayout` configuration is somehow invalid - this I think we can refer to as an internal server error.
Without manual intervention the operations controller would always break here. This is why I think the discussion from above applies here and I think we should actually update the operation status in k8s and update the operation in Director, so that the client can have immediate feedback and we can get notified immediately (as it would take some time for us to have adequate alerts). Here's the discussion from above, which comments on this exact situation: https://github.com/kyma-incubator/compass/pull/1733#discussion_r582827243